### PR TITLE
Added metadata node to all NIRNodes

### PR DESCRIPTION
**Pull request by Jegp**
I added a field with metadata for all NIRNodes for #70 

Note also that this does *not* support nested structures. That's totally by design because I worry this grows out of hand. But it does support arrays.

### CODE REVIEW

**Comment by Jegp** [comment]
**Update**: The previous version used a feature ([`kw_only`](https://docs.python.org/3/library/dataclasses.html)) that was only supported in Python $\geq$ 3.10. I manually added the fields to all individual objects to avoid this. The downside is that we don't have the field in the abstract `NIRNode` class, but it should work on Python 3.9 which is EOL 2025.

File: nir/serialization.py [hunk 1]
```diff
@@ -4,46 +4,59 @@
...
 def try_byte_to_str(a: Union[bytes, Any]) -> Union[str, Any]:
     return a.decode("utf8") if isinstance(a, bytes) else a
 
 
-def read_node(node: Any) -> nir.typing.NIRNode:
+def read_node(node: Any) -> ir.NIRNode:
     """Read a graph from a HDF/conn5 file."""
     if node["type"][()] == b"Affine":
-        return nir.Affine(weight=node["weight"][()], bias=node["bias"][()])
+        return ir.Affine(
```
Comment (matjobst) [inline nir/serialization.py:25]: Why the change from nir.Affine to ir.Affine? All Node types are directly accessible as nir.Affine, nir.Conv2d etc.

File: tests/test_readwrite.py [hunk 2]
```diff
@@ -40,9 +47,29 @@ def factory_test_graph(ir: nir.NIRGraph):
...
+                assert np.array_equal(v, d2[k])
+            elif isinstance(v, bytes):
+                assert v.decode("utf8") == d2[k]
+            else:
+                assert v == d2[k]
+
+    metadata = {"some": "metadata", "with": 2, "data": np.array([1, 2, 3])}
+    node.metadata = metadata
+    compare_dicts(node.metadata, metadata)
+    tmp = tempfile.mktemp()
```
Comment (matjobst) [inline tests/test_readwrite.py:63]: tempfile.mktemp is actually marked deprecated. They recommend tempfile.mkstemp as more secure variant:
https://docs.python.org/3/library/tempfile.html#tempfile.mktemp

Furthermore, neither mktemp nor mkstemp automatically remove the file. 
I would instead recommend tempfile.TemporaryFile which can be used as context manager, see example here:
https://docs.python.org/3/library/tempfile.html#tempfile-examples
```Python
# create a temporary file using a context manager
with tempfile.TemporaryFile() as fp:
    fp.write(b'Hello world!')
    fp.seek(0)
    fp.read()
# file is now closed and removed
```

File: nir/serialization.py (hunk 1 above)
Comment (Jegp) [reply nir/serialization.py:25]: Good catch! It should indeed be accessed via `nir`

File: tests/test_readwrite.py (hunk 2 above)
Comment (Jegp) [reply tests/test_readwrite.py:63]: Thank you for the input. And, it appears that [h5py also supports temporary files/RawIOBase files as inputs](https://docs.h5py.org/en/stable/high/file.html#python-file-like-objects). I updated the test
